src/libtsduck/dtv/analysis/tsTr101Analyzer.cpp

```diff
@@ -0,0 +1,503 @@
+#include <tsArgs.h>
+#include <tsBinaryTable.h>
+#include <tsPAT.h>
+#include <tsPMT.h>
+#include <tsjsonObject.h>
+
+// Defined in TR 101 290 Section 5.2.1
+#define PAT_INTERVAL (500l * SYSTEM_CLOCK_FREQ / 1000)
+
+// Defined in TR 101 290 Section 5.2.1
+#define PMT_INTERVAL (500l * SYSTEM_CLOCK_FREQ / 1000)
+
+// To the best of my knowledge, there is no specification that defines how frequently the CAT should be transmitted.
+#define CAT_VALID_INTERVAL (10l * SYSTEM_CLOCK_FREQ)
+
+// Defined in TR 101 290 Section 5.2.2
+#define PTS_REPETITION_INTERVAL (700l * SYSTEM_CLOCK_FREQ / 1000)
+
+#define PCR_DISCONTINUITY_LIMIT (100l * SYSTEM_CLOCK_FREQ /  1000)
+#define PCR_REPETITION_LIMIT (100l * SYSTEM_CLOCK_FREQ /  1000)
+#define PCR_ACCURACY_LIMIT_DOUBLE (500.0 / 1e9)
+
+ts::TR101_290Analyzer::TR101_290Analyzer(DuckContext& duck): _duck(duck) {
+    auto ptr = std::make_shared<ts::TR101_290Analyzer::ServiceContext>(0, ServiceContext::ServiceContextType::Pat);
+    _services.insert(std::make_pair(0, ptr));
+
+    _demux.addPID(PID_PAT);
+    _demux.addPID(PID_CAT);
+}
+
+void ts::TR101_290Analyzer::handleTable(SectionDemux& demux, const BinaryTable& table)
+{
+    auto service = getService(table.sourcePID());
+
+    if (table.tableId() == TID_PAT && service->_type == ServiceContext::Pat) {
+        const auto pat = PAT(_duck, table);
+
+        // Assign PMTs.
+        for (auto [service_id, pid] : pat.pmts) {
+            auto s2 = getService(pid);
+            s2->_type = ServiceContext::Pmt;
+            s2->_pmt_service_id = service_id;
+            _demux.addPID(pid);
+        }
+
+        // Remove PMT assignments.
+        for (auto it = _services.begin(); it != _services.end(); ++it) {
+            if (it->second->_type == ServiceContext::Pmt && !pat.pmts.contains(it->second->_pmt_service_id)) {
+                it->second->_type = ServiceContext::Unassigned;
+                _demux.removePID(it->first);
+            }
+        }
+    } else if (table.tableId() == TID_PMT && service->_type == ServiceContext::Pmt) {
+        const auto pmt = PMT(_duck, table);
+
+        // Ensure all PIDs are assigned to this service.
+        for (auto it = pmt.streams.begin(); it != pmt.streams.end(); it++) {
+            const auto s2 = getService(it->first);
+            s2->_type = ServiceContext::Assigned;
+            s2->_pmt_service_id = service->_pmt_service_id;
+        }
+
+        // Remove PIDs that are Assigned to this service ID.
+        for (auto it = _services.begin(); it != _services.end(); ++it) {
+            if (it->second->_type == ServiceContext::Assigned && it->second->_pmt_service_id == service->_pmt_service_id) {
+                if (!pmt.streams.contains(it->first)) {
+                    it->second->_type = ServiceContext::Unassigned;
+                }
+            }
+        }
+    }
+}
+
+void ts::TR101_290Analyzer::handleSection(SectionDemux& demux, const Section& table)
+{
+    if (table.sectionNumber() != 0) {
```

Comment:
I hope this is what you meant when you said that I had to look out for section numbers.

Reply:
Yes. The table cycle is the distance between two sections with the same number (typically section #0), not the distance between two sections (for instance sections 0 and 1 of the same table).

Reply:
And, for the sake of readability, you should call your parameter `section`, not `table`  (not the same concept, one is a subcomponent of the other).